homeassistant/components/assist_pipeline/pipeline.py

```diff
@@ -393,6 +409,81 @@ def end(self) -> None:
             )
         )
 
+    async def prepare_wake_word_detection(self) -> None:
+        """Prepare wake-word-detection."""
+        # Need to add to pipeline store
+        engine = wake_word.async_default_engine(self.hass)
+        assert engine is not None, "No wake word engine"
+
+        wake_provider = wake_word.async_get_wake_word_detection_entity(
+            self.hass, engine
+        )
+        if wake_provider is None:
+            raise WakeWordDetectionError(
+                code="wake-provider-missing",
+                message=f"No wake-word-detection provider for: {engine}",
+            )
+
+        self.wake_engine = engine
+        self.wake_provider = wake_provider
+
+    async def wake_word_detection(
+        self,
+        stream: AsyncIterable[bytes],
+        audio_buffer: list[bytes],
+    ) -> wake_word.DetectionResult | None:
+        """Run wake-word-detection portion of pipeline. Returns detection result."""
+        self.process_event(
+            PipelineEvent(
+                PipelineEventType.WAKE_START,
+                {
+                    "engine": self.wake_engine,
+                },
+            )
+        )
+
+        chunk_buffer: deque[bytes] = deque(maxlen=5)
+
+        async def timestamped_stream() -> AsyncIterable[tuple[bytes, int]]:
+            """Yield audio with timestamps (milliseconds since start of stream)."""
+            timestamp = 0
+            async for chunk in stream:
+                yield chunk, timestamp
+                timestamp += (len(chunk) // 2) // 16  # milliseconds @ 16Khz
+                chunk_buffer.append(chunk)
+
+        try:
+            # Detect wake word(s)
+            result = await self.wake_provider.async_process_audio_stream(
+                timestamped_stream()
+            )
+        except Exception as src_error:
+            _LOGGER.exception("Unexpected error during wake-word-detection")
+            raise SpeechToTextError(
+                code="wake-stream-failed",
+                message="Unexpected error during wake-word-detection",
+            ) from src_error
+
+        _LOGGER.debug("wake-word-detection result %s", result)
+
+        if result is None:
+            wake_output: dict[str, Any] = {"detected": False}
+        else:
+            audio_buffer.extend(chunk_buffer)
+            wake_output = {
+                "detected": True,
```

Comment:
Is this necessary? If it’s not detected we will timeout and sent an error event.